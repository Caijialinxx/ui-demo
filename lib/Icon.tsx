import React from 'react';
import './importIcons.js';

interface IconProps {
  name: string;
}

const Icon: React.FunctionComponent<IconProps> = (props) => {
  return (
    <svg width="1em" height="1em">
      <use xlinkHref={`#${props.name}`}/>
    </svg>);
};

export default Icon;