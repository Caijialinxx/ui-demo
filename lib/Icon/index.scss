.cui-icon {
  width: 1em;
  height: 1em;
  &-spin { animation: iconSpin 1s infinite linear; }
  @keyframes iconSpin {
    from { transform: rotate(0); }
    to { transform: rotate(360deg); }
  }
  &-shake { animation: iconShake 0.6s linear; }
  @keyframes iconShake {
    from, to { transform: rotate(0); }
    20%, 80% { transform: rotate(15deg); }
    35%, 65% { transform: rotate(-15deg); }
    50% { transform: rotate(15deg); }
  }
  &-shake-horizontal { animation: iconShakeHorizontal 0.8s linear; }
  @keyframes iconShakeHorizontal {
    from, to { transform: translateX(0); }
    15%, 45%, 75% { transform: translateX(-3px); }
    30%, 60%, 90%  { transform: translateX(3px); }
  }
  &-shake-vertical { animation: iconShakeVertical 0.8s linear; }
  @keyframes iconShakeVertical {
    from, to { transform: translateY(0); }
    15%, 45%, 75% { transform: translateY(2px); }
    30%, 60%, 90%  { transform: translateY(-2px); }
  }
}