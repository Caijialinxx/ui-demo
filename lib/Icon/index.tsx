import React from 'react';
import '../importIcons.js';
import './index.scss';
import joinClasses from '../helpers/joinClasses';

interface IconProps extends React.SVGAttributes<SVGSVGElement | SVGUseElement> {
  name: string;
}

const Icon: React.FunctionComponent<IconProps> = ({name, className, ...restProps}) => {
  return (
    <svg className={joinClasses('cui-icon', className)} {...restProps}>
      <use xlinkHref={`#${name}`}/>
    </svg>);
};

export default Icon;