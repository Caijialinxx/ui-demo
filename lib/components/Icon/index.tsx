import React, {useEffect, useState} from 'react';
import './importIcons.js';
import './index.scss';
import joinClasses from '../../helpers/joinClasses';

interface IconProps extends React.SVGAttributes<SVGSVGElement | SVGUseElement> {
  name: string;
  animationType?: 'spin' | 'shake' | 'shake-horizontal' | 'shake-vertical';
  animationTrigger?: 'click' | 'hover' | 'auto';
  // TODO: 从 iconfont.cn 引入 icon
  // TODO: 自定义 IconComponent
}

const Icon: React.FunctionComponent<IconProps> = ({name, className, animationType, animationTrigger, ...restProps}) => {
  const [animationClassName, setAnimationClassName] = useState<string>('');
  const [animationEventHandler, setAnimationEventHandler] = useState({});
  useEffect(() => {
    if (animationType) {
      if (animationTrigger === 'hover') {
        const onMouseOver: React.MouseEventHandler<SVGSVGElement | SVGUseElement> = () => {
          setAnimationClassName('cui-icon-' + animationType);
        };
        const onMouseLeave: React.MouseEventHandler<SVGSVGElement | SVGUseElement> = () => {
          setAnimationClassName('');
        };
        setAnimationEventHandler({onMouseOver, onMouseLeave});
      } else if (animationTrigger === 'click') {
        const onClick: React.MouseEventHandler<SVGSVGElement | SVGUseElement> = () => {
          setAnimationClassName('cui-icon-' + animationType);
        };
        const onAnimationEnd: React.MouseEventHandler<SVGSVGElement | SVGUseElement> = () => {
          setAnimationClassName('');
        };
        setAnimationEventHandler({onClick, onAnimationEnd});
      } else {
        setAnimationClassName('cui-icon-' + animationType);
      }
    }
  }, []);
  return (
    <svg fill="currentColor" className={joinClasses('cui-icon', className,
      name === 'loading' && 'cui-icon-spin',
      animationClassName
    )} {...animationEventHandler} {...restProps}>
      <use xlinkHref={`#${name}`}/>
    </svg>);
};

Icon.defaultProps = {
  animationTrigger: 'auto'
};

export default Icon;