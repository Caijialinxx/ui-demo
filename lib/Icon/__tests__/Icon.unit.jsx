import React from 'react'
import renderer from 'react-test-renderer'
import Icon from '../index'

describe('Icon', () => {
  it('传name', () => {
    const json = renderer.create(<Icon name="emoji-cool"/>).toJSON()
    expect(json).toMatchSnapshot()
  })
  it('传name、className', () => {
    const json = renderer.create(<Icon name="emoji-happy" className="icon"/>).toJSON()
    expect(json).toMatchSnapshot()
  })
  it('传name、className、事件', () => {
    const json = renderer.create(<Icon name="emoji-happy" className="icon" onClick={() => console.log(1)}/>).toJSON()
    expect(json).toMatchSnapshot()
  })
})