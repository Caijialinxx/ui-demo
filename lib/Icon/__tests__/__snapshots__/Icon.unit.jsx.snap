// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon 传name 1`] = `
<svg
  className="cui-icon"
>
  <use
    xlinkHref="#emoji-cool"
  />
</svg>
`;

exports[`Icon 传name、className 1`] = `
<svg
  className="cui-icon icon"
>
  <use
    xlinkHref="#emoji-happy"
  />
</svg>
`;

exports[`Icon 传name、className、事件 1`] = `
<svg
  className="cui-icon icon"
  onClick={[Function]}
>
  <use
    xlinkHref="#emoji-happy"
  />
</svg>
`;
